src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,163 @@
+    {
+        return a.txout.nValue > b.txout.nValue;
+    }
+} descending;
+
+/*
+ * This is the Branch and Bound Coin Selection algorithm designed by Mark Erhardt. It is an exact match algorithm where the exact match
+ * is a range with the lower bound being the value we want to spend and the upper bound being that value plus the additional cost
+ * required to create and spend a change output. To do this, the algorithm builds a binary tree where each node is a UTXO and whether
+ * that UTXO is included or not in the current coin selection. The tree is searched in include-first order. Each UTXO is first included
+ * and the current selection is evaluated for whether it is within the target range. If it is over the threshold, we try excluding
+ * the UTXO previously added and the branch of the tree involving its inclusion is then not explored. This process is repeated until
+ * a solution is found (i.e. selected value falls within the range), the exhaustion limit is reached, or the tree is exhausted and no
+ * solution was found.
+ *
+ * To find the best possible solution, we use a waste metric. The waste metric is defined as the cost to spend the current inputs
+ * now minus the cost to spend the current inputs later, plus the amount exceeding the target value. We search the tree to find the
+ * set of UTXOs which falls within our range and minimizes waste.
+ *
+ * An additional optimization of this algorithm implemented here is a lookahead value which maintains the total value of the UTXO set
+ * of all unexplored UTXOs (i.e. UTXOs that have not yet been included or excluded). This allows us to cut a branch if the remaining
+ * amount is not sufficient to reach our target.
+ *
+ * SelectCoinsBnB Arguments:
+ * const std::vector<CInputCoin>& utxo_pool -> The set of UTXOs that we are choosing from. These UTXOs will be sorted in descending order
+ *                                             by effective value and the CInputCoins' values are their effective values.
+ * const CAmount& target_value -> This is the value that we want to select. It is the lower bound of the range.
+ * const CAmount& cost_of_change -> This is the cost of creating and spending a change output. This plus target_value is the upper bound
+ *                                  of the range.
+ * std::set<CInputCoin>& out_set -> This is an output parameter for the set of CInputCoins that have been selected.
+ * CAmount& value_ret -> This is an output parameter for the total value of the CInputCoins that were selected.
+ * CAmount& fee_ret -> This is an output parameter for the value of the transaction fees for the CInputCoins that were selected.
+ */
+
+static const int remaining_tries = 100000;
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount& fee_ret)
+{
+    out_set.clear();
+    CAmount value_track = 0;
+
+    if (utxo_pool.empty()) {
```

Comment:
I think this is redundant with https://github.com/bitcoin/bitcoin/pull/10637/commits/f41f751f2e3ad69470624a6f55db3c1683712182#diff-491a507b32a78f89c3e066deedc98171R72

perhaps move that check here, to avoid deferencencing an invalid iterator.